pkg/cache/context.go

```diff
@@ -527,12 +564,20 @@ func (ctx *Context) IsPodFitNode(name string, node string) error {
 func (ctx *Context) AssumePod(name string, node string) error {
 	ctx.lock.Lock()
 	defer ctx.lock.Unlock()
-
 	if pod, ok := ctx.schedulerCache.GetPod(name); ok {
 		// when add assumed pod, we make a copy of the pod to avoid
 		// modifying its original reference. otherwise, it may have
 		// race when some other go-routines accessing it in parallel.
 		assumedPod := pod.DeepCopy()
+		// assume pod volumes, this will update bindings info in cache
+		if ctx.volumeBinder != nil {
+			// assume pod volumes before assuming the pod
+			// this will update scheduler cache with essential PV/PVC binding info
+			if _, err := ctx.volumeBinder.Binder.AssumePodVolumes(assumedPod, node); err != nil {
+				return err
+			}
+		}
+
```

Comment:
If the node does not exist we will have assumed the volumes on a node that does not exists. This does not make sense. It will also not keep the copied pod object around as it should be garbage collected when it is not added to the cache.
I think this all needs to be inside the check for the `targetNode`

Reply:
If the node doesn't exist, then bindPodVolumes will fail. If that fails, this will trigger the release operation and by then forgetPod will be called. This should be consistent.
And for the copied pod, that is stored in `ctx.schedulerCache`, this is to avoid directly modifying the original reference of the pod.

Reply:
`AssumePod` in the context is called on reconcile from the core, the node comes from core. Scheduling should fail at that point as the core is using a node that is not in the shim. We should retry scheduling or fail, not proceed as if nothing happened and then hope the volume bind fails later. What happens if we do not have volumes? 

I really don't think we should even call `AssumePodVolumes` if we know the node does not exist.

We also call `AssumePodVolumes` on the deep copy, when the node is not found that deep copy of the object is **not** added to the cache. We assume the volumes on a pod that we immediately drop. Bind on the original object will fail as the volumes have never been assumed.

Relying on a failure of `BindPodVolumes` to forget the pod is strange as we have not assumed the pod and never will.

Reply:
Got the point now, yeah, we should check the node existence for assumePodVolumes too, to keep them in sync.